### ISSUE TITLE
USE-QA for answer retrieval, fine-tuning

### ISSUE THREAD

**Issue by setu4993** [tensorflow/hub#428]
I had two questions about using the USE-QA model.

1. On the TF hub page for the [model](https://tfhub.dev/google/universal-sentence-encoder-multilingual-qa/2), in the 'Intended Use' section, it says:
`It is trained on a variety of data sources and tasks, with the goal of learning text representations that are useful out-of-the-box to retrieve an answer given a question, as well as question and answers across different languages.`

However, I can't find a good example for _how_ to retrieve the answer given a sentence text.

2. I cannot set `trainable=True` for either of the QA model versions on TF Hub. Are there other ways I can fine-tune them? Also, will that fine-tune the underlying QA architecture?

**Comment by maximneumann** [comment]
To 1: Following the example on https://tfhub.dev/google/universal-sentence-encoder-multilingual-qa/2
it seems that you need to provide next to the questions as well potential answers, and the task is then to match the most likely answer based on the retrieved embeddings. 
The np.inner() provides this and could be extended to:
print("Most likely answer: ", resonses[np.argmax(np.inner(question_embeddings, response_embeddings))])

To 2: You could start with constructing a model using Keras in TF-2 with KerasLayer:
model = tf.keras.Sequential([
  hub.KerasLayer(hub_handle, input_shape=[], dtype=tf.string, trainable=True),
  ...])
model.build()
model.compile(....)
model.fit(...)

The details of the model depend on the specific task and the data you are using. 
More information might be available in the references or by contacting the actual authors of this model. 
Please see the References section at the end of the page: https://tfhub.dev/google/universal-sentence-encoder-multilingual-qa/2

**Comment by setu4993** [comment]
Thanks for the help, @maximneumann.

Not completely clear on 1 (because a lookup from a list is opposite to what the prose seems to suggest), but I think it makes more sense to contact the authors directly.

**Comment by PExplorer** [comment]
Hi @setu4993, Are you able to fine tune USE_QA for specific domain text?

**Comment by setu4993** [comment]
@PExplorer: I chose not to fine-tune USE-QA when I realized it is not trivially possible to also encode context alongwith the Q-A pairs for fine-tuning the whole model. Moreover, fine-tuning the base model wasn't the best idea since the number of training samples for my training task were quite limited.

Instead, I went the transfer learning way and built a model on top of the multilingual USE model that served as the QA model.

**Comment by PExplorer** [comment]
@setu4993 Thanks for your prompt reply. In my use case, we have sufficient number of data points for training USE-QA Model(passage,question,answer(part of passage)). Using USE, retrieval of passages based on similarity between question and answer is not that great. So I was keen on USE-QA for fine tuning on specific domain data. I am still figuring out the way to provide input dict to USE-QA where as for USE, it is straight forward. Please share if you had any thoughts on figuring out the input format or steps you have tried. 

Thanks again,

**Comment by setu4993** [comment]
@PExplorer: I haven't tried this, but a model like the one described above (reproduced partly below), followed by fully-connected layers and a dot-product similarity at the end to calculate the loss (with labels as one-hot encodings within the batch) should work.
```python
model = tf.keras.Sequential([
hub.KerasLayer(hub_handle, input_shape=[], dtype=tf.string, trainable=True),
...])
```

To input the response and context into the model, this could work:
```python
encodings = model.signatures['response_encoder'](input=tf.constant([sentences[0][0]]), context=tf.constant([sentences[0][1]]))
```
More detailed example in [this Colab](https://colab.research.google.com/github/tensorflow/hub/blob/master/examples/colab/retrieval_with_tf_hub_universal_encoder_qa.ipynb#scrollTo=FwDUryIfSLp2).
